Various changes and improvements.

jwt27 opened the pull request:
Hey there, I've been hacking your build scripts a bit. The most notable improvements in this branch are:

- No longer download and build mpc/mpfr/gmp libraries separately. Instead we can use a script `contrib/download_prerequisites` in the gcc source tree.
- Checkpoints added to build system, skipping over some parts which have already been built/configured. This allows a user to resume when errors occur, and also avoids rebuilding binutils when installing a different version of gcc.
- Modular build scripts. The entire build process is handled by `script/build.sh`. Version name scripts (`script/*.*.*`) now only contain some version-specific variables. This should make it easier to support new gcc versions when they are released, fix bugs, and add new features.
- New variables `MAKE_JOBS` and `GCC_CONFIGURE_OPTIONS` allow a user to run multi-threaded builds, and pass custom options to gcc's `./configure`.

Some additional testing will likely be required. For example I don't know if `tar -a` is supported on all platforms. There may be more issues I'm unaware of.

I've also made various other minor changes and fixes. Please see the commit log and let me know what you think.

jwt27 commented:
I just pushed some more updates. The build script now allows you to specify which components and versions to install or upgrade. I also added support for building gdb.  
So for example, to build gcc version 7.1 with the latest binutils and gdb, you would use:
`./build-djgpp.sh base binutils gcc-7.1.0 gdb`

To just install the latest version of everything, you can use:
`./build-djgpp.sh all`

Let me know what you think.

jwt27 commented:
I pushed another update on a new branch: [tar](https://github.com/jwt27/build-djgpp/tree/tar). This branch builds GNU tar if it's not installed, so we can unpack all files with `-a` and no longer need to patch `unpack-gcc.sh` for OSX.  
This will need to be tested by someone on OSX since I'm unable to build tar on my mingw system.

jwt27 committed:
set default djgpp TARGET to i386-pc-msdosdjgpp.

lhzdem commented:
谢谢

jwt27 committed:
Revert "fix creating manpages directory."

This reverts commit 93bfe4e07f1c5e4ce65d085895966fcf3d1c555e.

Revert "install info/man files in target directory too."

This reverts commit 648ba9d482f9453470d51b18ebc03321ecef7613.

jwt27 committed:
add shebang line in setenv script

jwt27 committed:
change tests builds to i386-pc-msdosdjgpp

jwt27 committed:
add link-i586 script for *-pc-msdosdjgpp targets
for compatibility with other distributions which are compiled with target=i586-pc-msdosdjgpp.

jwt27 committed:
fix link-i586 script

jwt27 committed:
update readme

jwt27 committed:
debian package system

jwt27 committed:
deploy ubuntu ppa.

jwt27 committed:
support loading git sources from tar file.

jwt27 committed:
pack git sources into tar files.

jwt27 committed:
fix package names in debian/control file

jwt27 committed:
release ppa for multiple ubuntu versions

jwt27 committed:
drop ppa support for xenial

jwt27 committed:
remove binutils-djgpp package for focal

jwt27 committed:
make gcc-djgpp in focal depend on 2.34-3ubuntu1+1.4.

jwt27 committed:
use empty DESTDIR for temporarily installed tools.

jwt27 committed:
simplify msys2 pacman command in readme

jwt27 committed:
run test builds every 14 days.

jwt27 committed:
download gcc dependencies separately only if --build-deb is given. use download_prerequisites script otherwise.

jwt27 committed:
add lto to default enable-languages option.

jwt27 committed:
add binutils 2.34 for djgpp

jwt27 committed:
mv ppa signing key to /tmp

jwt27 committed:
use if statement to suppress error on deploy

jwt27 committed:
add gcc 9.3.0

jwt27 committed:
run test builds once a month

jwt27 committed:
patch djgpp gcc 9.3 and binutils 2.34 to use "b" flag for named bss sections.
also move djgpp patches to subdirectories.

jwt27 committed:
update debian changelog

jwt27 committed:
select correct djgpp version when upgrading gcc

If djgpp-cvs is already installed, rebuild that version instead of 2.05.

jwt27 committed:
patch djgpp ldscript to link .bss.* back in .bss

jwt27 committed:
update readme

jwt27 committed:
drop PPA for disco, launchpad no longer accepts new packages for it

jwt27 committed:
add PPA for Ubuntu Groovy

jwt27 committed:
always rebuild libc when rebuilding gcc

jwt27 committed:
add gcc 10.1

jwt27 committed:
add option --no-git-reset

jwt27 committed:
add cmake in test build dependencies

jwt27 committed:
fix syntax error in sed expression

jwt27 committed:
update readme

jwt27 committed:
remove patch for gcc 10 which has been merged in djcross-gcc

jwt27 committed:
add gcc 10.2

jwt27 committed:
add gdb 9.2

jwt27 committed:
configure gdb 8 and earlier to use python2

jwt27 committed:
add gdb 10.1

jwt27 committed:
add binutils 2.35.1

jwt27 committed:
add binutils 2.35.1 for djgpp

jwt27 committed:
disable scheduled test builds

jwt27 committed:
remove build status icon

jwt27 committed:
update test build dependencies

jwt27 committed:
init gpg homedir before importing key

jwt27 committed:
drop ubuntu eoan build

jwt27 committed:
build ppa for ubuntu hirsute (fixes #19)

jwt27 committed:
remove package installation example for mingw32

jwt27 committed:
list python headers dependency for gdb

jwt27 committed:
add newlib 4.0.0 and 4.1.0

jwt27 committed:
add binutils 2.35.2 and 2.36

jwt27 committed:
add binutils 2.36.1

jwt27 committed:
build ppa for ubuntu impish (fixes #23)

jwt27 committed:
add binutils 2.37

jwt27 committed:
add gcc 10.3.0

jwt27 committed:
add gdb 10.2 and 11.1

jwt27 committed:
fix dead link

jwt27 committed:
drop ppa for ubuntu groovy

jwt27 committed:
build ppa for ubuntu jammy

jwt27 committed:
sync ia16 configure flags with tkchia/build-ia16 (fixes #24)

jwt27 committed:
run ubuntu test builds with gcc-11

jwt27 committed:
install help2man in test builds

jwt27 committed:
install python2-dev in test builds

jwt27 committed:
add binutils 2.38

jwt27 committed:
update pgp key

jwt27 committed:
drop PPA for ubuntu hirsute

jwt27 committed:
update readme

jwt27 committed:
add avr-libc 2.1.0, avrdude 6.4, avarice 2.14

jwt27 committed:
fix simulavr build

jwt27 committed:
fix simulavr parallel build

jwt27 committed:
fix avarice build

jwt27 committed:
add gcc 11.1.0, 11.2.0

jwt27 committed:
add Watt-32 package

jwt27 committed:
include watt32 in test builds

jwt27 committed:
add debian package for Watt-32

jwt27 committed:
install nasm for test builds

jwt27 committed:
rename directory for watt32 errno files

jwt27 committed:
patch watt32 makefile to build on windows

jwt27 committed:
allow empty CFLAGS_FOR_TARGET

jwt27 committed:
new and improved method to store installed package versions

jwt27 committed:
use forward slashes in gcc specs file

jwt27 committed:
remove specs file before building libc or watt32

jwt27 committed:
fix ownership and permissions of installed files

jwt27 committed:
add binutils as dependency for watt32

jwt27 committed:
silence identical file warnings

jwt27 committed:
update watt-32 build procedure

jwt27 committed:
no need to force a libc rebuild on compiler upgrade

jwt27 committed:
add 'watt32' to 'all'

jwt27 committed:
use octal umask

jwt27 committed:
patch std::errc to use Watt-32 error codes

jwt27 committed:
add gcc option -mno-watt to disable Watt-32 include path

jwt27 committed:
set WATT_ROOT in setenv.cmd with regular slash and no quotes

jwt27 committed:
reduce code duplication

jwt27 committed:
fix typo in specs file

jwt27 committed:
compile target libraries with -ffunction-sections -fdata-sections

jwt27 committed:
always rebuild watt32 tools on non-amd64 or non-linux platforms.

neta79 reviewed (commented):
Hi. Thank you for your continued effort into making this environment even more awesome.
Question: soo, I may have misinterpreted things, but I'm pretty sure this line of development (the commit I am attaching this feedback to) should be granted a public branch.
I was only able to track these new changes by lookig at the commit flow. It was an accidental discovery :)

This is super-useful! Please make it into its own publi branch!  Thank you :)

jwt27 commented:
Hi @neta79!

I'm still actively developing this fork over at https://github.com/jwt27/build-gcc.  I did change the name, maybe that makes it more difficult to find.  Most recently I added support for building Watt-32.

The whole toolchain also available as a Ubuntu package, for use in Github Actions:
https://launchpad.net/~jwt27/+archive/ubuntu/djgpp-toolchain

jwt27 committed:
remove gcc specs option (doesn't work)

jwt27 committed:
remove merged watt32 patch

jwt27 committed:
silence warning on Mac

jwt27 committed:
add gdb 11.2, 12.1

jwt27 committed:
build PPA for ubuntu kinetic [fixes #26]

jwt27 committed:
update readme

jwt27 committed:
update default gdb version

jwt27 committed:
add gcc 12.1.0

jwt27 committed:
drop --always-make flag for watt32 utils

jwt27 committed:
set pkg-config path in setenv script

jwt27 committed:
list M4 as dependency

Commit:
Update README.md

Add notation concerning needing dos2unix to build djgpp.

jwt27 committed:
use msys2 perl on mingw-w64 (fixes #27)

jwt27 committed:
list dos2unix dependency for fedora/msys2

jwt27 committed:
debian: use separate version numbers for binary packages

jwt27 committed:
drop PPA for ubuntu Impish

jwt27 committed:
correct bash path in debian/rules

jwt27 committed:
keep ubuntu release name in binary package version numbers

jwt27 committed:
make djgpp-cvs the default version

jwt27 committed:
enhance setenv script

jwt27 committed:
add gcc 12.2.0

jwt27 committed:
add binutils 2.39

jwt27 committed:
improve debian package system

jwt27 committed:
export {C,CXX}FLAGS_FOR_TARGET

jwt27 committed:
debian: configure gcc with --libexecdir=/usr/lib

jwt27 committed:
debian: add convenience package 'djgpp'

jwt27 committed:
debian: install documentation in exec-prefix

jwt27 committed:
debian: drop 'extra' packages and install their contents in exec-prefix

jwt27 committed:
debian: add convenience package 'djgpp-doc'

jwt27 committed:
add variable to pass options to all autotools configure scripts

jwt27 committed:
debian: set environment vars from dpgk-buildflags

jwt27 committed:
debian: remove hyphens from package version numbers

jwt27 committed:
prevent accidentally uploading unreleased packages

jwt27 committed:
debian: move meta-packages to new repo

jwt27 committed:
search for libraries in ${EXEC_PREFIX}/local/

jwt27 committed:
debian: remove pgp key before build

jwt27 committed:
debian: fix stupid build error

jwt27 committed:
debian: add virtual packages for gcc and libc

jwt27 committed:
debian: set architecture for doc packages to 'all'

jwt27 committed:
update go32 stub in binutils to match libc version

jwt27 committed:
actions: use checkout@v3

jwt27 committed:
build ppa for ubuntu lunar [fixes: #31]

jwt27 committed:
release debian package v9

jwt27 committed:
run test builds with gcc 12

jwt27 committed:
actions: use cache@v3

jwt27 committed:
use ccache in test builds

jwt27 committed:
actions: use GITHUB_PATH

jwt27 committed:
fix gcc specs file path

jwt27 committed:
actions: fix ccache dir

jwt27 committed:
fix stupid typo in readme

cwwcww committed:
Add missing nasm dependency

jwt27 committed:
mention dxe3gen 1.0.5

jwt27 committed:
list `which` as dependency on fedora [#33]

jwt27 committed:
add gdb 13.1

jwt27 committed:
add binutils 2.40

jwt27 committed:
list available packages if none are given, without counting options

Fixes: #34

jwt27 committed:
add newlib 4.2.0, 4.3.0

jwt27 committed:
add avrdude 7.0

jwt27 committed:
reword

jwt27 committed:
rename meta-package 'default' -> 'minimal'

jwt27 committed:
add 'all'/'minimal' packages for all targets

jwt27 committed:
build PPA for Ubuntu Mantic

fixes: #35

jwt27 committed:
release debian package v10

jwt27 committed:
update watt32 errno lists

jwt27 committed:
add -mtune=pentium in debian build

jwt27 committed:
install wrapper script for pkg-config

jwt27 committed:
mention libc patch

jwt27 committed:
ensure old specs file is removed before build

jwt27 committed:
remove unused var BUILD_DXEGEN

jwt27 committed:
build PPA for Ubuntu Noble

fixes: #38

jwt27 committed:
fix setting CXXFLAGS_FOR_TARGET

jwt27 committed:
add binutils 2.41

jwt27 committed:
release debian package v11

jwt27 committed:
fix building zic.exe

jwt27 committed:
drop PPA for Ubuntu Kinetic

jwt27 committed:
use new syntax for workflow outputs

jwt27 committed:
pull gcc dependencies from ftpmirror.gnu.org

jwt27 committed:
update gcc dependencies

jwt27 committed:
no need for a separate function here

jwt27 committed:
set 'Architecture: all' for library packages

jwt27 committed:
export flags variables

jwt27 committed:
define WIN32_LEAN_AND_MEAN on mingw

jwt27 committed:
define WIN32_LEAN_AND_MEAN only while compiling gcc

fixes: #41

jwt27 committed:
improve package list display, drop duplicates

jwt27 committed:
remove dummy binutils package files for djgpp

jwt27 committed:
add binutils 2.42

jwt27 committed:
add gdb 13.2, 14.1, 14.2

jwt27 committed:
install gmp/mpfr for test builds

jwt27 committed:
add gcc 13.1, 13.2, 14.1

jwt27 committed:
fix avarice build on recent gcc versions

patches adapted from:
https://github.com/avrdudes/avarice/commit/c339b0d7c1d0c98274be6ea6db93ad477dbfe530
https://github.com/avrdudes/avarice/commit/a02e296cedd48f56b4f7779abb0a04fb2c63cf87

jwt27 committed:
copy go32stub patch to newer binutils versions

jwt27 committed:
add gdb 15.1

jwt27 committed:
bump default gdb version

jwt27 committed:
fix gdb 8.2.1 build error (implicit declaration of wcwidth())

Patch imported from gdb 8.3.
upstream commit 16bfc2f9705b40a11052f465b83fea2ec0904ce5

ref: #43

jwt27 committed:
add binutils 2.43

jwt27 committed:
release debian package v12

jwt27 committed:
ubuntu PPA: drop lunar, drop mantic, add oracular

jwt27 committed:
bump sed version for macos

(But do we need this still?  Users can just install it via brew)

jwt27 committed:
use --with-build-sysroot when installing via DESTDIR

fixes: #44

jwt27 committed:
gdb patch: Do not assume ELF in dwarf2/read.c

Patch imported from gdb master.  This fixes a segfault when reading symbols
from coff-go32 executables.
upstream commit: 5d683ae3dadd78fa88c243310404480555555246

jwt27 committed:
add binutils 2.43.1

jwt27 committed:
set gdb 15.1 as default for djgpp

jwt27 committed:
add gcc 14.2.0

jwt27 committed:
fix gcc 14.2 patch

jwt27 committed:
release debian package v13

jwt27 committed:
update workflow actions

jwt27 committed:
add Build-Depends: libgmp-dev, libmpfr-dev

jwt27 committed:
build gmp/mpfr/mpc/isl libs separately

jwt27 committed:
CI: fix ccache

jwt27 committed:
CI: silence "already installed" warnings from brew

jwt27 committed:
CI: remove unused brew packages

jwt27 committed:
use package version files for static libs

jwt27 committed:
djgpp: drop libemu

Workaround for #46.

jwt27 committed:
CI: remove 'rebuild' workflow

This has never been used, and doing this in CI is impractical.

jwt27 committed:
add gdb 15.2

jwt27 committed:
djgpp: re-enable libemu build

And apply the patch to djgpp-2.05 too.
fixes: #46

jwt27 committed:
add gdb 16.1

jwt27 committed:
add gdb 16.2

jwt27 committed:
add binutils 2.44

jwt27 committed:
remove references to python2

It doesn't exist anymore, and is no longer needed now that gdb 16 works with
djgpp.

jwt27 committed:
add gdb 16.3

jwt27 committed:
add libemu.a back in debian package

jwt27 committed:
build PPA for ubuntu Plucky

jwt27 committed:
add go32stub patch for binutils 2.44

jwt27 committed:
release debian package v14

jwt27 committed:
add workaround to build gmp-6.3.0 with gcc 15

fixes: #50

jwt27 committed:
configure --with-system-zlib

The bundled zlib versions fail to build on MacOS with clang >= 18.

ref: https://gitlab.kitware.com/cmake/cmake/-/issues/25755

jwt27 committed:
add avr-libc 2.2.0, 2.2.1

jwt27 committed:
add avrdude 8.0, 8.1

jwt27 committed:
add binutils 2.45

jwt27 committed:
remove avrdude 8.0, 8.1

They changed their whole build system, will have to look into that.

jwt27 committed:
CI: use gcc-15

jwt27 committed:
CI: ignore random brew warnings

jwt27 committed:
CI: actually install gcc-15

jwt27 committed:
CI: apparently we don't yet have gcc-15

jwt27 committed:
silence 'make clean'

This makes way too much noise when the compiler is not yet installed.

ref: #51

jwt27 committed:
build djgpp hostbin utils with -std=gnu17

fixes: #51

jwt27 committed:
patch djasm to make it compile with gcc 15

Also applied to djgpp-cvs.
fixes: #51 (but better)

jwt27 committed:
Merge branch 'djgpp-ppa'

jwt27 committed:
fix debian package version

v14 was already released so binutils 2.45 should go in v15

jwt27 committed:
add binutils 2.46.0

jwt27 committed:
add gdb 17.1

jwt27 committed:
build PPA for ubuntu Questing, Resolute

and drop Oracular, Plucky (obsolete)

jwt27 committed:
release debian package v15

jwt27 committed:
update workflow actions

jwt27 committed:
save CI build artifacts

jwt27 committed:
CI: add workflow for static build

jwt27 committed:
CI: use script/ci/clean.sh

jwt27 committed:
CI: fix missing backslash
